lib/sites-base.js

```diff
@@ -1,214 +1,308 @@
         }
       }
-
-      // Trusts the extension because it was already in uploadfs. -Tom
-      function getTempPath(file) {
-        return to.attachments.uploadfs.getTempPath() + '/' + require('path') + self.apos.utils.generateId() + require('path').extname(file);
+    };
+  
+    self.addTask('remove-orphans', 'Remove orphaned sites (db exists, but no matching site\ndoc). These will be temporary sites that did not get cleaned up. They\nnormally do not have any media in uploadfs, but if they do be aware it is\nnot cleaned up.', async function(apos, argv) {
+      const req = self.apos.tasks.getReq();
+      const dbs = await self.apos.db.admin().listDatabases();
+      const prefix = self.apos.shortName.replace(/dashboard$/, '');
+      // Make sure the dashboard is never a candidate to be removed
+      const names = dbs.databases.map(db => db.name).filter(db => db.startsWith(prefix) && (db !== `${prefix}dashboard`));
+      const known = await self.find(req, {}).trash(null).published(null).toArray().map(site => prefix + site._id);
+      const remove = _.difference(names, known);
+      for (const name of remove) {
+        const id = name.replace(prefix, '');
+        const base = prefix.replace(/-$/, '');
+        console.log(`dropping ${name}`);
+        await self.apos.db.db(name).dropDatabase(); 
       }
-    }
-  };
-
-  self.afterSave = async function(req, piece, options, callback) {
-
-    try {
-      if (req.newAdminPassword) {
-        await setAdminUser(req.newAdminPassword, piece, options);
-      }
-      return callback(null);
-    } catch (e) {
-      return callback(e);
-    }
-
-    async function setAdminUser(password, piece, options) {
-      const apos = await self.apos.options.multisite.getSiteApos(piece._id);
-      const req = apos.tasks.getReq();
-      const admin = await apos.users.find(req, { username: 'admin' }).toObject();
-      if (admin) {
-        admin.password = password;
-        console.log(admin);
-        return apos.users.update(req, admin);
-      } else {
-        const user = {
-          username: 'admin',
-          firstName: 'Admin',
-          lastName: 'User',
-          title: 'admin',
-          password
-        };
-        const group = await apos.groups.find(req, {
-          permissions: {
-            $in: [ 'admin' ]
-          }
-        }).toObject();
-        if (!group) {
-          group = await apos.groups.insert(req, {
-            title: 'admin',
-            permissions: [ 'admin' ]
-          });
+    });
+  
+    self.addTask('transition-shortname', 'Convert first element of hostname array to shortname in order to use baseUrlDomains.', async function(apos, argv) {
+      const req = self.apos.tasks.getReq();
+      const sites = await self.find(req, {}).toArray();
+      for(const site of sites) {
+        if (!site.shortName && site.hostnamesArray && site.hostnamesArray.length) {
+          site.shortName = site.hostnamesArray[0].hostname;
```

Comment:
This should be the *first part* of the hostname. That is, if the hostname is "foo.s", shortName should be "foo".